eth/stagedsync/bor_heimdall_shared.go

```diff
@@ -153,21 +133,37 @@ func fetchAndWriteHeimdallStateSyncEvents(
 	stateReceiverABI := cfg.stateReceiverABI
 	// Find out the latest eventId
 	var (
-		from uint64
-		to   time.Time
+		fromId uint64
+		to     time.Time
+		from   time.Time
 	)
 
 	blockNum := header.Number.Uint64()
 
+	if blockNum%config.CalculateSprintLength(blockNum) != 0 || blockNum == 0 {
+		// we fetch events only at beginning of each sprint
+		return lastStateSyncEventID, 0, 0, nil
+	}
+
+	prevHeader, err := blockReader.HeaderByNumber(ctx, tx, blockNum-config.CalculateSprintLength(blockNum))
+
+	if err != nil {
+		return lastStateSyncEventID, 0, time.Since(fetchStart), err
+	}
+
 	if config.IsIndore(blockNum) {
 		stateSyncDelay := config.CalculateStateSyncDelay(blockNum)
 		to = time.Unix(int64(header.Time-stateSyncDelay), 0)
+		from = time.Unix(int64(prevHeader.Time-stateSyncDelay), 0)
```

Comment:
can the new func `CalculateEventWIndow` you have added be used here as well instead of this if else?